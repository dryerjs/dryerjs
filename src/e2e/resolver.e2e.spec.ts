import { getIntrospectionQuery, GraphQLString } from 'graphql';
import {
    Arg,
    ExcludeOnCreate,
    NullableOnOutput,
    Omit,
    Partial,
    Pick,
    Property,
    Query,
    Required,
    RequiredOnCreate,
    Resolver,
} from 'dryerjs';
import { DryerTest } from './dryer-test';

export class Account {
    @Property()
    @ExcludeOnCreate()
    id: string;

    @Property()
    @RequiredOnCreate()
    email: string;

    @Property()
    @RequiredOnCreate()
    password: string;

    @Property()
    @NullableOnOutput()
    name: string;
}

class OmitCreateAccountInput extends Omit(Account, ['id']) {}

class PickCreateAccountInput extends Pick(Account, ['email', 'password', 'name']) {}

class PartialOfPickCreateAccountInput extends Partial(PickCreateAccountInput) {}

class RequiredOfPickCreateAccountInput extends Required(PickCreateAccountInput) {}

@Resolver(Account)
class AccountResolver {
    @Query(OmitCreateAccountInput)
    omit(@Arg('input') input: OmitCreateAccountInput) {
        return input;
    }

    @Query(GraphQLString)
    nullableInput(@Arg('email', false) email?: string) {
        return email || 'emtpy@email.com';
    }

    @Query(PickCreateAccountInput)
    pick(@Arg('input') input: PickCreateAccountInput) {
        return input;
    }

    @Query(PartialOfPickCreateAccountInput)
    partial(@Arg('input') input: PartialOfPickCreateAccountInput) {
        return input;
    }

    @Query(RequiredOfPickCreateAccountInput)
    require(@Arg('input') input: RequiredOfPickCreateAccountInput) {
        return input;
    }
}

export const dryer = DryerTest.init({
    modelDefinitions: [Account],
    resolvers: [AccountResolver],
});

describe('Resolver works', () => {
    beforeAll(async () => {
        await dryer.start();
    });

    it('Generate correct graphql schema', async () => {
        const query = getIntrospectionQuery();
        const response = await dryer.apolloServer.executeOperation({
            query,
        });
        const ignoreTypeNames = ['String', 'Int', 'Boolean'];
        const types = response.body['singleResult'].data.__schema.types.filter(({ name }) => {
            return !ignoreTypeNames.includes(name) && !name.startsWith('__');
        });
        expect(types).toMatchSnapshot();
    });

    it('nullableInput', async () => {
        const response1 = await dryer.makeSuccessRequest({
            query: `
                query {
                    nullableInput
                }
            `,
        });

        const response2 = await dryer.makeSuccessRequest({
            query: `
                query NullableInput($email: String) {
                    nullableInput(email: $email)
                }
            `,
            variables: { email: 'real@email.com' },
        });

        expect(response1.nullableInput).toEqual('emtpy@email.com');
        expect(response2.nullableInput).toEqual('real@email.com');
    });

    afterAll(async () => {
        await dryer.stop();
    });
});
