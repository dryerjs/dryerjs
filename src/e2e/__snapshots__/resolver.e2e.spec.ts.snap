// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Resolver works Generate correct graphql schema 1`] = `
[
  {
    "description": null,
    "enumValues": null,
    "fields": [
      {
        "args": [
          {
            "defaultValue": null,
            "description": null,
            "name": "id",
            "type": {
              "kind": "NON_NULL",
              "name": null,
              "ofType": {
                "kind": "SCALAR",
                "name": "String",
                "ofType": null,
              },
            },
          },
        ],
        "deprecationReason": null,
        "description": null,
        "isDeprecated": false,
        "name": "account",
        "type": {
          "kind": "NON_NULL",
          "name": null,
          "ofType": {
            "kind": "OBJECT",
            "name": "Account",
            "ofType": null,
          },
        },
      },
      {
        "args": [
          {
            "defaultValue": null,
            "description": null,
            "name": "options",
            "type": {
              "kind": "INPUT_OBJECT",
              "name": "PaginatedAccountsOptions",
              "ofType": null,
            },
          },
        ],
        "deprecationReason": null,
        "description": null,
        "isDeprecated": false,
        "name": "paginateAccounts",
        "type": {
          "kind": "OBJECT",
          "name": "PaginatedAccounts",
          "ofType": null,
        },
      },
      {
        "args": [],
        "deprecationReason": null,
        "description": null,
        "isDeprecated": false,
        "name": "allAccounts",
        "type": {
          "kind": "LIST",
          "name": null,
          "ofType": {
            "kind": "NON_NULL",
            "name": null,
            "ofType": {
              "kind": "OBJECT",
              "name": "Account",
              "ofType": null,
            },
          },
        },
      },
      {
        "args": [
          {
            "defaultValue": null,
            "description": null,
            "name": "input",
            "type": {
              "kind": "INPUT_OBJECT",
              "name": "CreateOmitCreateAccountInputInput",
              "ofType": null,
            },
          },
        ],
        "deprecationReason": null,
        "description": null,
        "isDeprecated": false,
        "name": "omit",
        "type": {
          "kind": "OBJECT",
          "name": "OmitCreateAccountInput",
          "ofType": null,
        },
      },
      {
        "args": [
          {
            "defaultValue": null,
            "description": null,
            "name": "email",
            "type": {
              "kind": "SCALAR",
              "name": "String",
              "ofType": null,
            },
          },
        ],
        "deprecationReason": null,
        "description": null,
        "isDeprecated": false,
        "name": "nullableInput",
        "type": {
          "kind": "SCALAR",
          "name": "String",
          "ofType": null,
        },
      },
      {
        "args": [
          {
            "defaultValue": null,
            "description": null,
            "name": "input",
            "type": {
              "kind": "INPUT_OBJECT",
              "name": "CreatePickCreateAccountInputInput",
              "ofType": null,
            },
          },
        ],
        "deprecationReason": null,
        "description": null,
        "isDeprecated": false,
        "name": "pick",
        "type": {
          "kind": "OBJECT",
          "name": "PickCreateAccountInput",
          "ofType": null,
        },
      },
      {
        "args": [
          {
            "defaultValue": null,
            "description": null,
            "name": "input",
            "type": {
              "kind": "INPUT_OBJECT",
              "name": "CreatePartialOfPickCreateAccountInputInput",
              "ofType": null,
            },
          },
        ],
        "deprecationReason": null,
        "description": null,
        "isDeprecated": false,
        "name": "partial",
        "type": {
          "kind": "OBJECT",
          "name": "PartialOfPickCreateAccountInput",
          "ofType": null,
        },
      },
      {
        "args": [
          {
            "defaultValue": null,
            "description": null,
            "name": "input",
            "type": {
              "kind": "INPUT_OBJECT",
              "name": "CreateRequiredOfPickCreateAccountInputInput",
              "ofType": null,
            },
          },
        ],
        "deprecationReason": null,
        "description": null,
        "isDeprecated": false,
        "name": "require",
        "type": {
          "kind": "OBJECT",
          "name": "RequiredOfPickCreateAccountInput",
          "ofType": null,
        },
      },
    ],
    "inputFields": null,
    "interfaces": [],
    "kind": "OBJECT",
    "name": "Query",
    "possibleTypes": null,
  },
  {
    "description": null,
    "enumValues": null,
    "fields": [
      {
        "args": [],
        "deprecationReason": null,
        "description": null,
        "isDeprecated": false,
        "name": "id",
        "type": {
          "kind": "NON_NULL",
          "name": null,
          "ofType": {
            "kind": "SCALAR",
            "name": "String",
            "ofType": null,
          },
        },
      },
      {
        "args": [],
        "deprecationReason": null,
        "description": null,
        "isDeprecated": false,
        "name": "email",
        "type": {
          "kind": "NON_NULL",
          "name": null,
          "ofType": {
            "kind": "SCALAR",
            "name": "String",
            "ofType": null,
          },
        },
      },
      {
        "args": [],
        "deprecationReason": null,
        "description": null,
        "isDeprecated": false,
        "name": "password",
        "type": {
          "kind": "NON_NULL",
          "name": null,
          "ofType": {
            "kind": "SCALAR",
            "name": "String",
            "ofType": null,
          },
        },
      },
      {
        "args": [],
        "deprecationReason": null,
        "description": null,
        "isDeprecated": false,
        "name": "name",
        "type": {
          "kind": "SCALAR",
          "name": "String",
          "ofType": null,
        },
      },
    ],
    "inputFields": null,
    "interfaces": [],
    "kind": "OBJECT",
    "name": "Account",
    "possibleTypes": null,
  },
  {
    "description": null,
    "enumValues": null,
    "fields": [
      {
        "args": [],
        "deprecationReason": null,
        "description": null,
        "isDeprecated": false,
        "name": "docs",
        "type": {
          "kind": "LIST",
          "name": null,
          "ofType": {
            "kind": "NON_NULL",
            "name": null,
            "ofType": {
              "kind": "OBJECT",
              "name": "Account",
              "ofType": null,
            },
          },
        },
      },
      {
        "args": [],
        "deprecationReason": null,
        "description": null,
        "isDeprecated": false,
        "name": "totalDocs",
        "type": {
          "kind": "SCALAR",
          "name": "Int",
          "ofType": null,
        },
      },
      {
        "args": [],
        "deprecationReason": null,
        "description": null,
        "isDeprecated": false,
        "name": "page",
        "type": {
          "kind": "SCALAR",
          "name": "Int",
          "ofType": null,
        },
      },
      {
        "args": [],
        "deprecationReason": null,
        "description": null,
        "isDeprecated": false,
        "name": "limit",
        "type": {
          "kind": "SCALAR",
          "name": "Int",
          "ofType": null,
        },
      },
      {
        "args": [],
        "deprecationReason": null,
        "description": null,
        "isDeprecated": false,
        "name": "hasPrevPage",
        "type": {
          "kind": "SCALAR",
          "name": "Boolean",
          "ofType": null,
        },
      },
      {
        "args": [],
        "deprecationReason": null,
        "description": null,
        "isDeprecated": false,
        "name": "hasNextPage",
        "type": {
          "kind": "SCALAR",
          "name": "Boolean",
          "ofType": null,
        },
      },
      {
        "args": [],
        "deprecationReason": null,
        "description": null,
        "isDeprecated": false,
        "name": "totalPages",
        "type": {
          "kind": "SCALAR",
          "name": "Int",
          "ofType": null,
        },
      },
    ],
    "inputFields": null,
    "interfaces": [],
    "kind": "OBJECT",
    "name": "PaginatedAccounts",
    "possibleTypes": null,
  },
  {
    "description": null,
    "enumValues": null,
    "fields": null,
    "inputFields": [
      {
        "defaultValue": null,
        "description": null,
        "name": "page",
        "type": {
          "kind": "SCALAR",
          "name": "Int",
          "ofType": null,
        },
      },
      {
        "defaultValue": null,
        "description": null,
        "name": "limit",
        "type": {
          "kind": "SCALAR",
          "name": "Int",
          "ofType": null,
        },
      },
    ],
    "interfaces": null,
    "kind": "INPUT_OBJECT",
    "name": "PaginatedAccountsOptions",
    "possibleTypes": null,
  },
  {
    "description": null,
    "enumValues": null,
    "fields": [
      {
        "args": [],
        "deprecationReason": null,
        "description": null,
        "isDeprecated": false,
        "name": "email",
        "type": {
          "kind": "NON_NULL",
          "name": null,
          "ofType": {
            "kind": "SCALAR",
            "name": "String",
            "ofType": null,
          },
        },
      },
      {
        "args": [],
        "deprecationReason": null,
        "description": null,
        "isDeprecated": false,
        "name": "password",
        "type": {
          "kind": "NON_NULL",
          "name": null,
          "ofType": {
            "kind": "SCALAR",
            "name": "String",
            "ofType": null,
          },
        },
      },
      {
        "args": [],
        "deprecationReason": null,
        "description": null,
        "isDeprecated": false,
        "name": "name",
        "type": {
          "kind": "SCALAR",
          "name": "String",
          "ofType": null,
        },
      },
    ],
    "inputFields": null,
    "interfaces": [],
    "kind": "OBJECT",
    "name": "OmitCreateAccountInput",
    "possibleTypes": null,
  },
  {
    "description": null,
    "enumValues": null,
    "fields": null,
    "inputFields": [
      {
        "defaultValue": null,
        "description": null,
        "name": "email",
        "type": {
          "kind": "NON_NULL",
          "name": null,
          "ofType": {
            "kind": "SCALAR",
            "name": "String",
            "ofType": null,
          },
        },
      },
      {
        "defaultValue": null,
        "description": null,
        "name": "password",
        "type": {
          "kind": "NON_NULL",
          "name": null,
          "ofType": {
            "kind": "SCALAR",
            "name": "String",
            "ofType": null,
          },
        },
      },
      {
        "defaultValue": null,
        "description": null,
        "name": "name",
        "type": {
          "kind": "SCALAR",
          "name": "String",
          "ofType": null,
        },
      },
    ],
    "interfaces": null,
    "kind": "INPUT_OBJECT",
    "name": "CreateOmitCreateAccountInputInput",
    "possibleTypes": null,
  },
  {
    "description": null,
    "enumValues": null,
    "fields": [
      {
        "args": [],
        "deprecationReason": null,
        "description": null,
        "isDeprecated": false,
        "name": "email",
        "type": {
          "kind": "NON_NULL",
          "name": null,
          "ofType": {
            "kind": "SCALAR",
            "name": "String",
            "ofType": null,
          },
        },
      },
      {
        "args": [],
        "deprecationReason": null,
        "description": null,
        "isDeprecated": false,
        "name": "password",
        "type": {
          "kind": "NON_NULL",
          "name": null,
          "ofType": {
            "kind": "SCALAR",
            "name": "String",
            "ofType": null,
          },
        },
      },
      {
        "args": [],
        "deprecationReason": null,
        "description": null,
        "isDeprecated": false,
        "name": "name",
        "type": {
          "kind": "SCALAR",
          "name": "String",
          "ofType": null,
        },
      },
    ],
    "inputFields": null,
    "interfaces": [],
    "kind": "OBJECT",
    "name": "PickCreateAccountInput",
    "possibleTypes": null,
  },
  {
    "description": null,
    "enumValues": null,
    "fields": null,
    "inputFields": [
      {
        "defaultValue": null,
        "description": null,
        "name": "email",
        "type": {
          "kind": "NON_NULL",
          "name": null,
          "ofType": {
            "kind": "SCALAR",
            "name": "String",
            "ofType": null,
          },
        },
      },
      {
        "defaultValue": null,
        "description": null,
        "name": "password",
        "type": {
          "kind": "NON_NULL",
          "name": null,
          "ofType": {
            "kind": "SCALAR",
            "name": "String",
            "ofType": null,
          },
        },
      },
      {
        "defaultValue": null,
        "description": null,
        "name": "name",
        "type": {
          "kind": "SCALAR",
          "name": "String",
          "ofType": null,
        },
      },
    ],
    "interfaces": null,
    "kind": "INPUT_OBJECT",
    "name": "CreatePickCreateAccountInputInput",
    "possibleTypes": null,
  },
  {
    "description": null,
    "enumValues": null,
    "fields": [
      {
        "args": [],
        "deprecationReason": null,
        "description": null,
        "isDeprecated": false,
        "name": "email",
        "type": {
          "kind": "SCALAR",
          "name": "String",
          "ofType": null,
        },
      },
      {
        "args": [],
        "deprecationReason": null,
        "description": null,
        "isDeprecated": false,
        "name": "password",
        "type": {
          "kind": "SCALAR",
          "name": "String",
          "ofType": null,
        },
      },
      {
        "args": [],
        "deprecationReason": null,
        "description": null,
        "isDeprecated": false,
        "name": "name",
        "type": {
          "kind": "SCALAR",
          "name": "String",
          "ofType": null,
        },
      },
    ],
    "inputFields": null,
    "interfaces": [],
    "kind": "OBJECT",
    "name": "PartialOfPickCreateAccountInput",
    "possibleTypes": null,
  },
  {
    "description": null,
    "enumValues": null,
    "fields": null,
    "inputFields": [
      {
        "defaultValue": null,
        "description": null,
        "name": "email",
        "type": {
          "kind": "SCALAR",
          "name": "String",
          "ofType": null,
        },
      },
      {
        "defaultValue": null,
        "description": null,
        "name": "password",
        "type": {
          "kind": "SCALAR",
          "name": "String",
          "ofType": null,
        },
      },
      {
        "defaultValue": null,
        "description": null,
        "name": "name",
        "type": {
          "kind": "SCALAR",
          "name": "String",
          "ofType": null,
        },
      },
    ],
    "interfaces": null,
    "kind": "INPUT_OBJECT",
    "name": "CreatePartialOfPickCreateAccountInputInput",
    "possibleTypes": null,
  },
  {
    "description": null,
    "enumValues": null,
    "fields": [
      {
        "args": [],
        "deprecationReason": null,
        "description": null,
        "isDeprecated": false,
        "name": "email",
        "type": {
          "kind": "NON_NULL",
          "name": null,
          "ofType": {
            "kind": "SCALAR",
            "name": "String",
            "ofType": null,
          },
        },
      },
      {
        "args": [],
        "deprecationReason": null,
        "description": null,
        "isDeprecated": false,
        "name": "password",
        "type": {
          "kind": "NON_NULL",
          "name": null,
          "ofType": {
            "kind": "SCALAR",
            "name": "String",
            "ofType": null,
          },
        },
      },
      {
        "args": [],
        "deprecationReason": null,
        "description": null,
        "isDeprecated": false,
        "name": "name",
        "type": {
          "kind": "NON_NULL",
          "name": null,
          "ofType": {
            "kind": "SCALAR",
            "name": "String",
            "ofType": null,
          },
        },
      },
    ],
    "inputFields": null,
    "interfaces": [],
    "kind": "OBJECT",
    "name": "RequiredOfPickCreateAccountInput",
    "possibleTypes": null,
  },
  {
    "description": null,
    "enumValues": null,
    "fields": null,
    "inputFields": [
      {
        "defaultValue": null,
        "description": null,
        "name": "email",
        "type": {
          "kind": "NON_NULL",
          "name": null,
          "ofType": {
            "kind": "SCALAR",
            "name": "String",
            "ofType": null,
          },
        },
      },
      {
        "defaultValue": null,
        "description": null,
        "name": "password",
        "type": {
          "kind": "NON_NULL",
          "name": null,
          "ofType": {
            "kind": "SCALAR",
            "name": "String",
            "ofType": null,
          },
        },
      },
      {
        "defaultValue": null,
        "description": null,
        "name": "name",
        "type": {
          "kind": "NON_NULL",
          "name": null,
          "ofType": {
            "kind": "SCALAR",
            "name": "String",
            "ofType": null,
          },
        },
      },
    ],
    "interfaces": null,
    "kind": "INPUT_OBJECT",
    "name": "CreateRequiredOfPickCreateAccountInputInput",
    "possibleTypes": null,
  },
  {
    "description": null,
    "enumValues": null,
    "fields": [
      {
        "args": [
          {
            "defaultValue": null,
            "description": null,
            "name": "input",
            "type": {
              "kind": "NON_NULL",
              "name": null,
              "ofType": {
                "kind": "INPUT_OBJECT",
                "name": "CreateAccountInput",
                "ofType": null,
              },
            },
          },
        ],
        "deprecationReason": null,
        "description": null,
        "isDeprecated": false,
        "name": "createAccount",
        "type": {
          "kind": "NON_NULL",
          "name": null,
          "ofType": {
            "kind": "OBJECT",
            "name": "Account",
            "ofType": null,
          },
        },
      },
      {
        "args": [
          {
            "defaultValue": null,
            "description": null,
            "name": "input",
            "type": {
              "kind": "NON_NULL",
              "name": null,
              "ofType": {
                "kind": "INPUT_OBJECT",
                "name": "UpdateAccountInput",
                "ofType": null,
              },
            },
          },
        ],
        "deprecationReason": null,
        "description": null,
        "isDeprecated": false,
        "name": "updateAccount",
        "type": {
          "kind": "NON_NULL",
          "name": null,
          "ofType": {
            "kind": "OBJECT",
            "name": "Account",
            "ofType": null,
          },
        },
      },
      {
        "args": [
          {
            "defaultValue": null,
            "description": null,
            "name": "id",
            "type": {
              "kind": "NON_NULL",
              "name": null,
              "ofType": {
                "kind": "SCALAR",
                "name": "String",
                "ofType": null,
              },
            },
          },
        ],
        "deprecationReason": null,
        "description": null,
        "isDeprecated": false,
        "name": "deleteAccount",
        "type": {
          "kind": "OBJECT",
          "name": "DeleteResponse",
          "ofType": null,
        },
      },
    ],
    "inputFields": null,
    "interfaces": [],
    "kind": "OBJECT",
    "name": "Mutation",
    "possibleTypes": null,
  },
  {
    "description": null,
    "enumValues": null,
    "fields": null,
    "inputFields": [
      {
        "defaultValue": null,
        "description": null,
        "name": "email",
        "type": {
          "kind": "NON_NULL",
          "name": null,
          "ofType": {
            "kind": "SCALAR",
            "name": "String",
            "ofType": null,
          },
        },
      },
      {
        "defaultValue": null,
        "description": null,
        "name": "password",
        "type": {
          "kind": "NON_NULL",
          "name": null,
          "ofType": {
            "kind": "SCALAR",
            "name": "String",
            "ofType": null,
          },
        },
      },
      {
        "defaultValue": null,
        "description": null,
        "name": "name",
        "type": {
          "kind": "SCALAR",
          "name": "String",
          "ofType": null,
        },
      },
    ],
    "interfaces": null,
    "kind": "INPUT_OBJECT",
    "name": "CreateAccountInput",
    "possibleTypes": null,
  },
  {
    "description": null,
    "enumValues": null,
    "fields": null,
    "inputFields": [
      {
        "defaultValue": null,
        "description": null,
        "name": "id",
        "type": {
          "kind": "SCALAR",
          "name": "String",
          "ofType": null,
        },
      },
      {
        "defaultValue": null,
        "description": null,
        "name": "email",
        "type": {
          "kind": "SCALAR",
          "name": "String",
          "ofType": null,
        },
      },
      {
        "defaultValue": null,
        "description": null,
        "name": "password",
        "type": {
          "kind": "SCALAR",
          "name": "String",
          "ofType": null,
        },
      },
      {
        "defaultValue": null,
        "description": null,
        "name": "name",
        "type": {
          "kind": "SCALAR",
          "name": "String",
          "ofType": null,
        },
      },
    ],
    "interfaces": null,
    "kind": "INPUT_OBJECT",
    "name": "UpdateAccountInput",
    "possibleTypes": null,
  },
  {
    "description": null,
    "enumValues": null,
    "fields": [
      {
        "args": [],
        "deprecationReason": null,
        "description": null,
        "isDeprecated": false,
        "name": "deleted",
        "type": {
          "kind": "NON_NULL",
          "name": null,
          "ofType": {
            "kind": "SCALAR",
            "name": "Boolean",
            "ofType": null,
          },
        },
      },
      {
        "args": [],
        "deprecationReason": null,
        "description": null,
        "isDeprecated": false,
        "name": "id",
        "type": {
          "kind": "NON_NULL",
          "name": null,
          "ofType": {
            "kind": "SCALAR",
            "name": "String",
            "ofType": null,
          },
        },
      },
    ],
    "inputFields": null,
    "interfaces": [],
    "kind": "OBJECT",
    "name": "DeleteResponse",
    "possibleTypes": null,
  },
]
`;
